clash-lib/src/Clash/Driver/Manifest.hs

```diff
@@ -0,0 +1,269 @@
+{-# LANGUAGE LambdaCase #-}
+{-|
+Functions to read, write, and handle manifest files.
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Clash.Driver.Manifest where
+
+import           Control.Exception (tryJust)
+import           Control.Monad (guard)
+import qualified Crypto.Hash.SHA256 as Sha256
+import qualified Data.ByteString.Lazy as ByteStringLazy
+import           Data.ByteString (ByteString)
+import           Data.Hashable (hash)
+import           Data.Maybe (catMaybes)
+import qualified Data.Text as Text
+import qualified Data.Text.IO as Text
+import           Data.Text (Text)
+import           Data.Time (UTCTime)
+import qualified Data.Set as Set
+import           Text.Read (readMaybe)
+import           System.IO.Error (isDoesNotExistError)
+import           System.FilePath (takeDirectory, (</>))
+import           System.Directory (listDirectory, doesFileExist)
+
+import           Clash.Annotations.TopEntity.Extra ()
+import           Clash.Driver.Types
+import           Clash.Primitives.Types
+import           Clash.Core.Var (Id)
+import           Clash.Netlist.Types (TopEntityT, Component(..))
+import qualified Clash.Netlist.Types as Netlist
+import qualified Clash.Netlist.Id as Id
+import           Clash.Primitives.Util (hashCompiledPrimMap)
+import           Clash.Util.Graph (callGraphBindings)
+
+import           Util (OverridingBool(Auto))
+
+-- | Information about the generated HDL between (sub)runs of the compiler
+data Manifest
+  = Manifest
+  { manifestHash :: Int
+    -- ^ Hash of the TopEntity and all its dependencies
+  , successFlags  :: (Int,Int,Bool)
+    -- ^ Compiler flags used to achieve successful compilation:
+    --
+    --   * opt_inlineLimit
+    --   * opt_specLimit
+    --   * opt_floatSupport
+  , portInNames  :: [Text]
+  , portOutNames :: [Text]
+  , componentNames :: [Text]
+    -- ^ Names of all the generated components for the @TopEntity@ (does not
+    -- include the names of the components of the @TestBench@ accompanying
+    -- the @TopEntity@).
+  , topComponent :: Text
+    -- ^ Design entry point. This is usually the component annotated with a
+    -- @TopEntity@ annotation.
+  , fileNames :: [(FilePath, ByteString)]
```

Comment:
Would a `HashMap` be a better choice of type here?

Reply:
But there's nothing that would benefit from log(n) complexity, is there?

Reply:
I know neither if there would be a benefit now, or if there would be a benefit for any future plans @martijnbastiaan has vis-a-vis improving this area of the compiler

Reply:
I don't have any immediate plans, but it would be nice if we `manifest` would obsolete `edam.py`. With that in mind, we would probably expand the properties in `fileNames` (which should really be `files`?). A hashmap would make it easier to work with in, say, Python.

Reply:
I plan on making another PR to make the manifest file a JSON. I'll put these considerations in there.